src/flash.rs

```diff
@@ -2,6 +2,294 @@
+            Ok(())
+        }
+    }
+
+    /// Erase sector which contains start_offset
+    pub fn page_erase(&mut self, start_offset: u32) -> Result<()> {
+        self.valid_address(start_offset)?;
+
+        // Unlock Flash
+        self.unlock()?;
+
+        // Set Page Erase
+        self.flash.cr.cr().modify(|_, w| w.per().set_bit());
+
+        // Write address bits
+        // NOTE(unsafe) This sets the page address in the Address Register.
+        // The side-effect of this write is that the page will be erased when we
+        // set the STRT bit in the CR below. The address is validated by the
+        // call to self.valid_address() above.
+        unsafe {
+            self.flash
+                .ar
+                .ar()
+                .write(|w| w.far().bits(FLASH_START + start_offset));
+        }
+
+        // Start Operation
+        self.flash.cr.cr().modify(|_, w| w.strt().set_bit());
+
+        // Wait for operation to finish
+        while self.flash.sr.sr().read().bsy().bit_is_set() {}
+
+        // Check for errors
+        let sr = self.flash.sr.sr().read();
+
+        // Remove Page Erase Operation bit
+        self.flash.cr.cr().modify(|_, w| w.per().clear_bit());
+
+        // Re-lock flash
+        self.lock()?;
+
+        if sr.wrprterr().bit_is_set() {
+            self.flash.sr.sr().modify(|_, w| w.wrprterr().clear_bit());
+            Err(Error::EraseError)
+        } else {
+            if self.verify {
+                // By subtracting 1 from the sector size and masking with
+                // start_offset, we make 'start' point to the beginning of the
+                // page, and similarly 'end' at the end of the page. We do this
+                // because the entire page should have been erased, regardless
+                // of where in the page the given 'start_offset' was.
+                let start = start_offset & !(self.sector_sz.kbytes() as u32 - 1);
+                let end = start + self.sector_sz.kbytes() as u32 - 1;
+                for idx in start..end {
+                    let write_address = (FLASH_START + idx as u32) as *const u16;
+                    let verify: u16 = unsafe { core::ptr::read_volatile(write_address) };
+                    if verify != 0xFFFF {
+                        return Err(Error::VerifyError);
+                    }
+                }
+            }
+
+            Ok(())
+        }
+    }
+
+    /// Erase the Flash Sectors from FLASH_START + start_offset to length
+    pub fn erase(&mut self, start_offset: u32, length: usize) -> Result<()> {
+        self.valid_length(start_offset, length)?;
+
+        // Erase every sector touched by start_offset + length
+        for offset in
+            (start_offset..start_offset + length as u32).step_by(self.sector_sz.kbytes() as usize)
+        {
+            self.page_erase(offset)?;
+        }
+
+        // Report Success
+        Ok(())
+    }
+
+    /// Retrieve an slice of data from FLASH_START +offset
```

Comment:
Nitpick: should be `a slice`. Also, would be nice to have the math in a code block, and a space after the `+`